platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/AccessorsClassPath.kt

```diff
@@ -283,8 +283,38 @@ internal
 fun importsRequiredBy(candidateTypes: List<TypeAccessibility>): List<String> =
     defaultPackageTypesIn(
         candidateTypes
-            .filterIsInstance<TypeAccessibility.Accessible>()
-            .map { it.type.kotlinString }
+            .filterIsInstance<TypeAccessibility.Accessible>().let { accessibleTypes ->
+                val annotations = object {
+                    var typeNames: MutableSet<String>? = null
+
+                    fun addTypeName(typeName: String) {
+                        if (typeNames == null) {
+                            typeNames = mutableSetOf()
+                        }
+                        typeNames!!.add(typeName)
+                    }
+
+                    fun visitAnnotationValue(annotationValueRepresentation: AnnotationValueRepresentation) {
+                        when (annotationValueRepresentation) {
+                            is AnnotationValueRepresentation.PrimitiveValue,
+                            is AnnotationValueRepresentation.ValueArray -> Unit
+                            is AnnotationValueRepresentation.AnnotationValue -> visitAnnotation(annotationValueRepresentation.representation)
+                            is AnnotationValueRepresentation.EnumValue -> addTypeName(annotationValueRepresentation.type.kotlinString)
+                            is AnnotationValueRepresentation.ClassValue -> addTypeName(annotationValueRepresentation.type.kotlinString)
+                        }
+                    }
+
+                    fun visitAnnotation(annotation: AnnotationRepresentation) {
+                        addTypeName(annotation.type.kotlinString)
+                        annotation.values.values.forEach { annotationValue -> visitAnnotationValue(annotationValue) }
+                    }
+                }
+
+                accessibleTypes.forEach { accessibleType -> accessibleType.optInRequirements.forEach { annotations.visitAnnotation(it) } }
+
+                val typeNames = accessibleTypes.map { it.type.kotlinString }
+                if (annotations.typeNames == null) typeNames else typeNames + annotations.typeNames!!.toList()
+            }
```

Comment:
💅 Let's extract explaining function(s) to make this more readable and make the intent clearer when reading

Reply:
Done in [`94b4038` (#33746)](https://github.com/gradle/gradle/pull/33746/commits/94b4038c04cf5524cf1b3a9702cbfe900d0f9526)